test/cloudtest/test_storaged_shared_fate.py

```diff
@@ -0,0 +1,219 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import subprocess
+import time
+from textwrap import dedent
+
+from materialize.cloudtest.application import MaterializeApplication
+
+CLUSTER_SIZE = 8
+NUM_SOURCES = 4
+
+
+def populate(mz: MaterializeApplication, seed: int) -> None:
+    create_sources = "\n".join(
+        f"""
+            > CREATE SOURCE s{i}
+              IN CLUSTER storaged_shared_fate
+              FROM KAFKA CONNECTION kafka (TOPIC 'testdrive-storaged-shared-fate-${{testdrive.seed}}')
+              FORMAT BYTES
+              ENVELOPE NONE;
+    """
+        for i in range(NUM_SOURCES)
+    )
+
+    create_views = "\n".join(
+        f"""
+            > CREATE MATERIALIZED VIEW v{i} AS SELECT COUNT(*) FROM t1 UNION ALL SELECT COUNT(*) FROM s{i};
+    """
+        for i in range(NUM_SOURCES)
+    )
+
+    create_indexes_and_counts = "\n".join(
+        f"""
+            > CREATE DEFAULT INDEX ON v{i};
+
+            > SELECT COUNT(*) > 0 FROM s{i};
+            true
+    """
+        for i in range(NUM_SOURCES)
+    )
+
+    mz.testdrive.run(
+        # TODO: I need both clusters? Can't create index/mv in cluster which has sources connected with IN CLUSTER
+        input=dedent(
+            f"""
+            > CREATE CLUSTER storaged_shared_fate REPLICAS (storaged_shared_fate_replica (SIZE '{CLUSTER_SIZE}-1'));
+
+            > CREATE CLUSTER shared_fate REPLICAS (shared_fate_replica (SIZE '{CLUSTER_SIZE}-1'));
+
+            > SET cluster = storaged_shared_fate;
+
+            > CREATE CONNECTION kafka TO KAFKA (BROKER '${{testdrive.kafka-addr}}')
+
+            $ kafka-create-topic topic=storaged-shared-fate partitions={CLUSTER_SIZE}
+
+            $ kafka-ingest format=bytes topic=storaged-shared-fate repeat=1000
```

Comment:
Unfortunately turns out that using `FORMAT BYTES` does not load all the partitions :-( If we select from the `_progress` source, which provides information as to how many messages came from which partition:

```
materialize=> select * from s0_progress;
 partition | offset 
-----------+--------
 (7,)      | 0
 [0,0]     | 0
 [1,1]     | 999
 [2,2]     | 0
 [3,3]     | 0
 [4,4]     | 2
 [5,5]     | 0
 [6,6]     | 999
 [7,7]     | 1000
(9 rows)
```

Not ideal. Several Kafka partitions were empty meaning that any computed processes that happened to read from them would do no actual work.

How exactly Kafka distributes incoming messages to partitions is a bit of a black box (why is there a partition with 2 messages above?), but the easiest way out is to make sure that each kafka message also has a **key** and not just a value. Like this:
```
diff --git a/test/cloudtest/test_storaged_shared_fate.py b/test/cloudtest/test_storaged_shared_fate.py
index 0e1bdc3f6..68b59e527 100644
--- a/test/cloudtest/test_storaged_shared_fate.py
+++ b/test/cloudtest/test_storaged_shared_fate.py
@@ -60,8 +60,8 @@ def populate(mz: MaterializeApplication, seed: int) -> None:
 
             $ kafka-create-topic topic=storaged-shared-fate partitions={CLUSTER_SIZE}
 
-            $ kafka-ingest format=bytes topic=storaged-shared-fate repeat=1000
-            CDE${{kafka-ingest.iteration}}
+            $ kafka-ingest format=bytes key-format=bytes key-terminator=: topic=storaged-shared-fate repeat=1000
+            CDE${{kafka-ingest.iteration}}:CDE${{kafka-ingest.iteration}}
 
             {create_sources}
 
@@ -77,8 +77,8 @@ def populate(mz: MaterializeApplication, seed: int) -> None:
 
             > SET cluster = storaged_shared_fate;
 
-            $ kafka-ingest format=bytes topic=storaged-shared-fate repeat=1000
-            DEF${{kafka-ingest.iteration}}
+            $ kafka-ingest key-format=bytes format=bytes key-terminator=: topic=storaged-shared-fate repeat=1000
+            DEF${{kafka-ingest.iteration}}:DEF${{kafka-ingest.iteration}}
 
             > SET cluster = shared_fate;
 
@@ -109,16 +109,12 @@ def validate(mz: MaterializeApplication, seed: int) -> None:
 
             > INSERT INTO t1 SELECT 345000 + generate_series FROM generate_series(1, 1000);
 
-            $ kafka-ingest format=bytes topic=storaged-shared-fate repeat=1000
-            EFG${{kafka-ingest.iteration}}
+            $ kafka-ingest key-format=bytes format=bytes key-terminator=: topic=storaged-shared-fate repeat=1000
+            EFG${{kafka-ingest.iteration}}:EFG${{kafka-ingest.iteration}}
```

Once you do this, the distribution of messages becomes much more equitable:

```
materialize=> select * from s2_progress;
 partition | offset 
-----------+--------
 (7,)      | 0
 [0,0]     | 256
 [1,1]     | 262
 [2,2]     | 251
 [3,3]     | 267
 [4,4]     | 240
 [5,5]     | 242
 [6,6]     | 235
 [7,7]     | 247
```